### PR TITLE
fix: Move temporary files outside of the cache directory

### PR DESCRIPTION
**Pull request by kitadai31**
fix for #2120

The "ephemeral" directory is moved from cacheDir to filesDir.
The input.apk (exists if selected from storage) is moved to the ephemeral directory.

I'm not sure if the changes in `Session.kt` are right.
This is intended to delete input.apk as soon as possible when it is no longer needed.
If the apk is a downloaded apk or selected from installed apps, nothing happens on that line.

### CODE REVIEW

**Comment by oSumAtrIX** [comment]
Does this PR also delete the temporal files after they are not necessary anymore?

**Comment by kitadai31** [comment]
Yes
After patching is finished, only the output patched apk will be in the temporary folder
This apk will also be deleted when I back to the dashboard

**Comment by oSumAtrIX** [comment]
Which line of code deletes the temp files?

**Comment by kitadai31** [comment]
`File(cacheDir, "input.apk").delete()` removes a input apk which is selected from storage

Other temporary files are already deleted after patching since before this PR

File: app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt [hunk 1]
```diff
@@ -16,7 +16,7 @@ class Filesystem(private val app: Application) {
      * A directory that gets cleared when the app restarts.
      * Do not store paths to this directory in a parcel.
      */
-    val tempDir = app.cacheDir.resolve("ephemeral").apply {
+    val tempDir = app.filesDir.resolve("ephemeral").apply {
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt:19]: Use getDir instead

File: app/src/main/java/app/revanced/manager/patcher/Session.kt [hunk 2]
```diff
@@ -129,6 +129,7 @@ class Session(
 
     override fun close() {
         tempDir.deleteRecursively()
+        File(cacheDir, "input.apk").delete()
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/patcher/Session.kt:132]: PatcherViewModel already deletes it.

File: app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt [hunk 3]
```diff
@@ -17,15 +18,19 @@ import kotlinx.coroutines.Dispatchers
...
 class AppSelectorViewModel(
     private val app: Application,
     private val pm: PM,
     private val patchBundleRepository: PatchBundleRepository
-) : ViewModel() {
-    private val inputFile = File(app.cacheDir, "input.apk").also {
+) : ViewModel(), KoinComponent {
+    private val fs: Filesystem by inject()
+
+    private val inputFile = File(fs.tempDir, "input.apk").also {
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:33]: tempDir is only valid for the duration of the app process, but this needs to live longer than that. Use getDir instead.

**Review by Axelen123 — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt (hunk 1 above)
Comment (kitadai31) [reply app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt:19]: Since getDir creates the directory when it is called, I thought it was not suitable because I would have to create the directory myself.
I noticed that getDir and mkdirs() have different permissions on the folders created.
Should I still use getDir?

**Comment by kitadai31** [comment]
Reflected the review, thank you.

Also, I moved the deleting selected apk code to PatcherWorker.

Currently, PatcherViewModel deletes the selected apk on `onCleared()` method.
But this may cause the selected apk (input.apk) to remain in internal storage because onCleared() won't be called if the users exit the task from recent apps.
If user exits the task after patching has finished without returning to the dashboard by accident, the input.apk remains in the app data.
Therefore, I think the input.apk should be deleted as soon as possible when patching is ended.
This is the reason I moved the code.

Is this OK?

File: app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt [hunk 4]
```diff
@@ -25,7 +26,7 @@ class AppSelectorViewModel(
     private val pm: PM,
     private val patchBundleRepository: PatchBundleRepository
 ) : ViewModel() {
-    private val inputFile = File(app.cacheDir, "input.apk").also {
+    private val inputFile = File(app.getDir("ephemeral", Context.MODE_PRIVATE), "input.apk").also {
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: "ephemeral" is deleted by `Filesystem` every time the app process is restarted. Use another directory.

File: app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt (hunk 1 above)
Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/data/platform/Filesystem.kt]: Yes, because that is what the android docs recommends.

File: app/src/main/java/app/revanced/manager/patcher/worker/PatcherWorker.kt [hunk 5]
```diff
@@ -193,6 +193,9 @@ class PatcherWorker(
             Result.failure()
         } finally {
             patchedApk.delete()
+            if (args.input is SelectedApp.Local && args.input.temporary) {
+                args.input.file.delete()
+            }
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/patcher/worker/PatcherWorker.kt:198]: What is the reason for this change?

**Review by Axelen123 — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt (hunk 4 above)
Comment (kitadai31) [reply app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: input.apk should be deleted every time the app restarted, shouldn't it?
Is input.apk needed after restarted?

File: app/src/main/java/app/revanced/manager/patcher/worker/PatcherWorker.kt (hunk 5 above)
Comment (kitadai31) [reply app/src/main/java/app/revanced/manager/patcher/worker/PatcherWorker.kt:198]: Please read the message above 🙇‍♂️

Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/patcher/worker/PatcherWorker.kt:198]: My bad, makes sense.

File: app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt (hunk 4 above)
Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: We cannot delete the input APK every time the app process is restarted because of [system-initiated process death](https://galex.dev/posts/process-death-is-the-rule-not-the-exception/#users-experience-system-initiated-process-death-a-lot). The process is killed when the app is in the background, but when the user navigates back a new process is launched and state is restored. System-initiated process death would definitely cause problems if the app is killed while the user is in a screen between `AppSelectorScreen` and `PatcherScreen`.

Comment (kitadai31) [reply app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: OK understood.
I now get what you mean by "process".

Since this is a single file, I think the destination will be `filesDir`. Is that correct?
( `File(app.filesDir, "input.apk")` )

Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: Use something like `getDir("app_selector", Context.MODE_PRIVATE)`. I remember having problems when trying to use `filesDir` directly.

Comment (kitadai31) [reply app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: Isn't it [`dataDir`](https://developer.android.com/reference/android/content/Context#getDataDir()) ?

> Apps should not use this path directly; they should instead use getFilesDir(), getCacheDir(), getDir(java.lang.String, int), or other storage APIs on this class. 

`filesDir` is a standard location to place a file
https://developer.android.com/training/data-storage/app-specific#internal-access-store-files
I replaced it with `File(app.filesDir, "input.apk")` and there were no problems.

Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/viewmodel/AppSelectorViewModel.kt:29]: In that case `filesDir` is fine.

**Review by Axelen123 — APPROVED** [review]
